secret_store/src/key_server_cluster/io/handshake.rs

```diff
@@ -0,0 +1,320 @@
+					return Ok((stream, Err(Error::InvalidNodeId)).into());
+				}
+
+				self.other_node_id = Some(message.node_id.into());
+				self.other_confirmation_plain = Some(message.confirmation_plain.into());
+				if self.is_active {
+					self.shared_key = match compute_shared_key(self.self_key_pair.secret(),
+						self.other_node_id.as_ref().expect("filled couple of lines above; qed")
+					) {
+						Ok(shared_key) => Some(shared_key),
+						Err(err) => return Ok((stream, Err(err)).into()),
+					};
+
+					let message = match Handshake::<A>::make_private_key_signature_message(
+						self.self_key_pair.secret(),
+						self.other_confirmation_plain.as_ref().expect("filled couple of lines above; qed")
+					) {
+						Ok(message) => message,
+						Err(err) => return Ok((stream, Err(err)).into()),
+					};
+					(HandshakeState::SendPrivateKeySignature(write_encrypted_message(stream,
+						self.shared_key.as_ref().expect("filled couple of lines above; qed"),
+					message)), Async::NotReady)
+				} else {
+					let message = match Handshake::<A>::make_public_key_message(self.self_key_pair.public().clone(), self.self_confirmation_plain.clone()) {
+						Ok(message) => message,
+						Err(err) => return Ok((stream, Err(err)).into()),
+					};
+					(HandshakeState::SendPublicKey(write_message(stream, message)), Async::NotReady)
+				}
+			},
+			HandshakeState::SendPrivateKeySignature(ref mut future) => {
+				let (stream, _) = try_ready!(future.poll());
+
+				(HandshakeState::ReceivePrivateKeySignature(
+					read_message(stream)
+				), Async::NotReady)
+			},
+			HandshakeState::ReceivePrivateKeySignature(ref mut future) => {
+				let (stream, message) = try_ready!(future.poll());
+
+				let message = match message {
+					Ok(message) => match message {
+						Message::Cluster(ClusterMessage::NodePrivateKeySignature(message)) => message,
+						_ => return Ok((stream, Err(Error::InvalidMessage)).into()),
+					},
+					Err(err) => return Ok((stream, Err(err.into())).into()),
+				};
+
+				let other_node_public = self.other_node_id.as_ref().expect("other_node_id is filled in ReceivePublicKey; ReceivePrivateKeySignature follows ReceivePublicKey; qed");
+				if !verify_public(other_node_public, &*message.confirmation_signed, &self.self_confirmation_plain).unwrap_or(false) {
+					return Ok((stream, Err(Error::InvalidMessage)).into());
+				}
+
+				(HandshakeState::Finished, Async::Ready((stream, Ok(HandshakeResult {
+					node_id: self.other_node_id.expect("other_node_id is filled in ReceivePublicKey; ReceivePrivateKeySignature follows ReceivePublicKey; qed"),
+					shared_key: self.shared_key.clone().expect("shared_key is filled in Send/ReceivePublicKey; ReceivePrivateKeySignature follows Send/ReceivePublicKey; qed"),
+				}))))
+			},
+			HandshakeState::Finished => panic!("poll Handshake after it's done"),
+		};
+
+		self.state = next;
+		match result {
+			// by polling again, we register new future
+			Async::NotReady => self.poll(),
```

Comment:
this won't lead to unbound recursion if `state(n) = poll(state(n-1))`

Without this additional `self.poll()` this future would never be polled again after returning `Async::NotReady`. And as you noticed `try_ready!` prevents `self.poll()` to be called more than once for the same state 😄